chore(docs): add docs to create a new app route (#6204)
artsy/eigen: docs/adding_a_new_route.md

## Adding a new route

### Route in Eigen

> In Eigen, as in many native app, we have mainly two types for routes.

Reviewer: ```suggestion
In Eigen, as in many mobile app, we have mainly two types for routes.
```
is that what you meant 🤔?

- Global app routes
- Context specific app routes

### Global App Routes

> Global app routes in Eigen, are routes that accessible everywhere in the app using the `src/lib/navigation/navigate` helper.

Reviewer: ```suggestion
Global app routes in Eigen, are routes that are accessible everywhere in the app using the `src/lib/navigation/navigate` helper.
```

> Before you maing the decision of creating a new global route - Make sure it is really needed. In many occasions, there is no need for it and using an independent Navigation container is all you need.

Reviewer: ```suggestion
Before making the decision of creating a new global route - make sure it is really needed. In many occasions, there is no need for it and using an independent Navigation container is all you need.
```

   ```xml
   ...
   <data android:pathPrefix="/my-screen" />
   ...
   ```
3. Register your screen at `AppRegistry.tsx`.
   ```typescript
   MyScreen: reactModule(
      MyScreen, // The screen name in the app
      {
        // Show the screen as a modal always
        // PS: The screen will show up as a modal but won't be treated as modal. This is great to avoid common modal pifalls
        alwaysPresentModally?: boolean
        // This screen refers to a tab in the app
        isRootViewForTabName: BottomTabType,
        fullBleed?: boolean
        // Hide the close button inside the moda
        hasOwnModalCloseButton?: boolean
        hidesBackButton?: boolean
        // Sepecify how you want the screen to show up
        // See https://developer.apple.com/documentation/uikit/uiviewcontroller/1621355-modalpresentationstyle?language=objc
        modalPresentationStyle?: "fullScreen" | "pageSheet" | "formSheet"
        // If this module should only be shown in one particular tab, add it here
        onlyShowInTabName?: BottomTabType
      },
    ),
   ```

### Context specific app routes

> Context specific app routes in Eigen are routes that are part of an independent [NavigationContainer](https://reactnavigation.org/docs/navigation-container/). All routes using defined inside a `NavigationContainer` context will receive a [navigation](https://reactnavigation.org/docs/navigation-prop/) and [route](https://reactnavigation.org/docs/glossary-of-terms/#route-prop) props which allow you to navigate through the different screens in a stack, listen to navigation events and also send params.

Reviewer: ```suggestion
Context specific app routes in Eigen are routes that are part of an independent [NavigationContainer](https://reactnavigation.org/docs/navigation-container/). All routes defined inside a `NavigationContainer` context will receive a [navigation](https://reactnavigation.org/docs/navigation-prop/) and [route](https://reactnavigation.org/docs/glossary-of-terms/#route-prop) props which allow you to navigate through the different screens in a stack, listen to navigation events and also send params.
```

> Example of situations where to use an independent navigation container

Reviewer: ```suggestion
Example of situations where we should use an independent navigation container
```

> - You are adding a new screen within a route that can be only accessed through a link at that route and is only app available and it doesn't make sense to deep link the users into it or navigate through it from somewhere else.

Reviewer: should we split this one in a few bullet points with one example each? my brain exploded with this sentence 😅.

Author: makes sense, it's time to use the https://hemingwayapp.com/ again

export const MyScreenNavigationStack = () => {
  return (
    <NavigationContainer independent>
      <MyScreenStackNavigator.Navigator
        headerMode="screen"
        screenOptions={{
          headerShown: false, // hide or show navigation header
          // Add required screenOptions here. See https://reactnavigation.org/docs/group/#screenoptions
        }}
      >
        <MyScreenStackNavigator.Screen name="MyScreenFirstPage" component={MyScreenFirstPage} />
        <MyScreenStackNavigator.Screen name="MyScreenSecondPage" component={MyScreenSecondPage} />
        <MyScreenStackNavigator.Screen name="MyScreenThirdPage" component={MyScreenThirdPage} />
      </MyScreenStackNavigator.Navigator>
    </NavigationContainer>
  )
}
```

All screens inside the stack will now receive the `navigation` and `route` prop. For typescript support, you can define your route props as follows:

```typescript
interface MyScreenFirstPageProps
  extends StackScreenProps<
    MyScreenNavigationStack,
    "MyScreenFirstPage"
  > {
    // other props the page has
  }

const MyScreenFirstPage: React.FC<MyScreenFirstPageProps> = ({navigation, route}) => {
  ...
}
```

> or you can as well get the same values using `useNavigation` and `useRoute` props.

Reviewer: should we flip these, and put hooks first, "or", and then put the props option?

just so that we signal our preference to the hooks way.

Author: good point, this will save us from prop drilling. I will update
